CMakeLists.txt

```diff
@@ -0,0 +1,255 @@
+# integer (MPI) library written entirely in C.
+#
+
+cmake_minimum_required(VERSION 3.7)
+project(tommath VERSION 1.2.0)
+
+#-----------------------------------------------------------------------------
+# Include cmake modules
+#-----------------------------------------------------------------------------
+include(GNUInstallDirs)
+include(CMakePackageConfigHelpers)
+include(sources.cmake)
+
+# The only direct cmake argument for now
+option(BUILD_SHARED_LIBS "Build shared library and only the shared library if \"ON\", default is static" OFF)
+
+#-----------------------------------------------------------------------------
+# Compose CFLAGS
+#-----------------------------------------------------------------------------
+
+# Some information copied from makefile_include.mk
+
+# Basic set
+set(LTM_C_FLAGS  -Wall -Wsign-compare -Wextra -Wshadow)
+set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wdeclaration-after-statement -Wbad-function-cast -Wcast-align)
+set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wstrict-prototypes -Wpointer-arith)
+
+# Do we run the sanitizer?
+if(DEFINED ENV{SANITIZER})
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -fsanitize=undefined -fno-sanitize-recover=all -fno-sanitize=float-divide-by-zero)
+endif()
+
+if(DEFINED ENV{CONV_WARNINGS})
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -std=c89 -Wconversion -Wsign-conversion)
+    if($ENV{CONV_WARNINGS} EQUAL "strict")
+        set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wc++-compat)
+    endif()
+else()
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wsystem-headers)
+endif()
+
+if(DEFINED ENV{COMPILE_DEBUG})
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -g3)
+endif()
+
+if(DEFINED ENV{COMPILE_SIZE})
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Os)
+else()
+    if(NOT DEFINED ENV{IGNORE_SPEED})
+        set(LTM_C_FLAGS  ${LTM_C_FLAGS} -O3 -funroll-loops)
+        #x86 optimizations [should be valid for any GCC install though]
+        set(LTM_C_FLAGS  ${LTM_C_FLAGS} -fomit-frame-pointer)
+    endif()
+    # TODO:
+    # if(DEFINED ENV{COMPILE_LTO})
+    #     set(LTM_C_FLAGS  "${LTM_C_FLAGS} -flto")
+    #     set(LTM_LD_FLAGS "${LTM_LD_FLAGS} -flto")
+    #     #AR = $(subst clang,llvm-ar,$(subst gcc,gcc-ar,$(CC)))
+    # endif()
+endif()
+
+# TODO
+# Are the coming three checks really the best way?
+
+# What compiler do we have and what are their...uhm... peculiarities
+# TODO: is the check for a C++ compiler necessary?
+if( (CMAKE_C_COMPILER_ID MATCHES "(C|c?)lang") OR (CMAKE_CXX_COMPILER_ID MATCHES "(C|c?)lang"))
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wno-typedef-redefinition -Wno-tautological-compare -Wno-builtin-requires-header)
+endif()
+
+if( (CMAKE_C_COMPILER_ID MATCHES "mingw") OR (CMAKE_CXX_COMPILER_ID MATCHES "mingw"))
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-shadow)
+endif()
+
+if(DEFINED ENV{PLATFORM})
+    if($ENV{PLATFORM} MATCHES "Darwin")
+        set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-nullability-completeness)
+    endif()
+    if($ENV{PLATFORM} MATCHES "CYGWIN")
+        set(LTM_C_FLAGS  ${LTM_C_FLAGS} -no-undefined)
+    endif()
+endif()
```

Comment:
Is this really the best way to do this?

Reply:
I don't have the faintest idea although my bet is on: probably not. I could only spend about an hour on it before my coffee run out.

RegExp put in compiler checks because it seems as if `CMAKE_C_COMPILER_ID` contains either `Clang` or `clang` under certain circumstances (athough [CMake docs](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html) list only `Clang` ) and I found no way to make `MATCH` case insensitive.

There are several CMake build-ins for platform checks, see e.g. [this link](https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/How-To-Write-Platform-Checks) that has a couple of lists at the end of the page with the macros but I found no *exact* definitions for them (but did not dive into CMake's sourcecode).